web/Services/PostcodeService.cs

```diff
@@ -0,0 +1,116 @@
+    {
+        var postcodeResponseList = new List<PostcodeioResponseModel>();
+        var postcodesToLookupChunks = postcodes.Distinct()
+            .Where(p => p != "" && _dbContext.Postcodes.All(dbRecord => dbRecord.Postcode != p))
+            .Chunk(100);
+
+        var addPostcodeChunkToResponseList = postcodesToLookupChunks.Select(async p => postcodeResponseList.AddRange(await BulkGetLatLonForPostcodes(p.ToList())));
+        await Task.WhenAll(addPostcodeChunkToResponseList);
+        
+        var postcodeGroups = postcodeResponseList
+                .GroupBy(p => p.Result?.Lat == null || p.Result.Lon == null || p.Result.Postcode == null)
+                .ToDictionary(grouping => grouping.Key ? "invalid" : "valid", grouping => grouping.ToArray());
+    
+        if (postcodeGroups.ContainsKey("valid"))
+        {
+            AddPostcodesToPostcodeDb(postcodeGroups["valid"]);
+        }
+        if (postcodeGroups.ContainsKey("invalid") && postcodeGroups["invalid"].Length > 0)
+        {
+            _logger.LogWarning($"{postcodeGroups["invalid"].Length} postcodes cannot be converted to coordinates.");
+        }
+    }
+
+    private async Task<List<PostcodeioResponseModel>> BulkGetLatLonForPostcodes(List<string> postcodesToLookup)
+    {
+        var bulkResponseBody = await MakeBulkApiRequestToPostcodeApi(postcodesToLookup);
+        var postcodeResponse = TurnBulkResponseBodyToModel(bulkResponseBody);
+        return postcodeResponse;
+    }
+
+    private async Task<string> MakeBulkApiRequestToPostcodeApi(List<string> postcodes)
+    {
+        var postcodesJson = new { postcodes = postcodes.ToArray() };
+        var uri = "https://api.postcodes.io/postcodes/";
+        var bulkResponseBody = string.Empty;
+        _logger.LogInformation("Making API Request to {Uri}", uri);
+        try
+        {
+            var response = await _client.PostAsJsonAsync(uri, postcodesJson);
+            response.EnsureSuccessStatusCode();
+            bulkResponseBody = await response.Content.ReadAsStringAsync();
+            _logger.LogInformation("Successful API request to {Uri}", uri);
+        }
+        catch (Exception e)
+        {
+            _logger.LogWarning("Message :{Message} ", e.Message);
+        }
+
+        return bulkResponseBody;
+    }
+
+    private List<PostcodeioResponseModel> TurnBulkResponseBodyToModel(string bulkResponseBody)
+    {
+        var postcodeResponseList = new List<PostcodeioResponseModel>();
+        if (string.IsNullOrEmpty(bulkResponseBody))
+        {
+            return postcodeResponseList;
+        }
+        var jObjectResponse = JObject.Parse(bulkResponseBody);
+        var postcodeList = jObjectResponse["result"]!.ToList();
+        foreach (var postcode in postcodeList)
+        {
+            var postcodeDbModel = postcode.ToObject<PostcodeioResponseModel>();
+            postcodeResponseList.Add(postcodeDbModel!);
+        }
+        return postcodeResponseList;

```

Comment:
We can actually simplify this down a fair amount :). If you go through and make all of the method argument and return types IEnumerable instead of Lists, then this is what we get
```suggestion
    {
        if (string.IsNullOrEmpty(bulkResponseBody))
        {
            return new List<PostcodeioResponseModel>();
        }
        var jObjectResponse = JObject.Parse(bulkResponseBody);
        var postcodeList = jObjectResponse["result"]!;
        return postcodeList.Select(postcode => postcode.ToObject<PostcodeioResponseModel>()!);

```

In general it is better to use IEnumerable as often as possible instead of List, this is because as we are going through building our enumerable we are basically creating an order of operations, so starting with a, then perform method 1, then perform method 2, etc. When we call ToList it runs through and iterates and performs all these methods in actuality. The reason we want to call this ToList as late as possible, is that the black box of C#, when it compiles, it can simplify and speed up all the different methods we are asking it to perform on an enumerable before making it a list. This means that we can have a situation where instead of doing steps 1 and 2, then going to list, then doing steps 3 and 4 and doing to list, c# can merge steps 2 and 3 into one quicker step 5, and do steps 1 5 and 4 and then do a ToList.

TLDR, use IEnumerable and avoid ToList wherever possible (but don't worry about it too much as its a performance thing and this is not going to be a service that suffers as a result of very slightly less performant code)